File: src/0015_three_sum.py
"""
3 Sum

Given an integer array nums,
return all the triplets
    nums[i], nums[j], nums[k]]
such that i != j, i != k, and j != k,
and nums[i] + nums[j] + nums[k] == 0.

Notice that the solution set must not contain duplicate triplets.

Constraints:
    -) 3 <= nums.length <= 3000
    -) -10^5 <= nums[i] <= 10^5
"""

tests = (
    ([-1, 0, 1, 2, -1, -4], [[-1, -1, 2], [-1, 0, 1]]),
    ([0, 1, 1], []),
    ([0, 0, 0], [[0, 0, 0]]),
    ([-2, 0, 0, 2, 2], [[-2, 0, 2]]),
)


def three_sums(nums):
    """
    Idea: So hard ...
    Time: O(n log n) + O(n^2) -> O(n^2)
    Memo: O(1)
    """
    res = []
    nums.sort()
    for i, n in enumerate(nums):
        if i > 0 and n == nums[i - 1]:
            # required to eliminate duplicates [-1,-1,0,1]
            continue
        l, r = i + 1, len(nums) - 1
        while l < r:
            total = n + nums[l] + nums[r]
            if total == 0:
                res.append([n, nums[l], nums[r]])
                l += 1
                r -= 1
                while nums[l] == nums[l - 1] and l < r:
                    # required to eliminate duplicates
                    # [-2, 0, 0, 2, 2]
                    l += 1
            elif total < 0:
                l += 1
            else:
                r -= 1
    return res


if __name__ == "__main__":
    for nums, exp in tests:
        act = three_sums(nums)
        assert act == exp, f"act={act} != exp={exp}"

File: go/src/0015_three_sum.go
/*
Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] where nums[i] + nums[j] + nums[k] == 0, and the indices i, j and k are all distinct.

The output should not contain any duplicate triplets. You may return the output and the triplets in any order.

Example 1:
Input: nums = [-1,0,1,2,-1,-4]

Output: [[-1,-1,2],[-1,0,1]]
Explanation:
nums[0] + nums[1] + nums[2] = (-1) + 0 + 1 = 0.
nums[1] + nums[2] + nums[4] = 0 + 1 + (-1) = 0.
nums[0] + nums[3] + nums[4] = (-1) + 2 + (-1) = 0.
The distinct triplets are [-1,0,1] and [-1,-1,2].

Example 2:
Input: nums = [0,1,1]

Output: []
Explanation: The only possible triplet does not sum up to 0.

Example 3:
Input: nums = [0,0,0]

Output: [[0,0,0]]
Explanation: The only possible triplet sums up to 0.

Constraints:

3 <= nums.length <= 1000
-10^5 <= nums[i] <= 10^5


Recommended Time & Space Complexity
You should aim for a solution with O(n^2) time and O(1) space, where n is the size of the input array.
*/

package main

import (
	"fmt"
	"sort"
)

func threeSum(nums []int) [][]int {
	/*
	Time O(n^3)
	Space O(m) where m is the number of triplets
	*/
	resMap := map[[3]int]struct{}{}
	sort.Ints(nums)
	for i:=0; i<len(nums) - 2; i++ {
		for j:=i+1; j < len(nums) - 1; j++{
			for k:=j; k < len(nums); k++ {
				if nums[i] + nums[j] + nums[k] == 0 {
					resMap[[3]int{nums[i], nums[j], nums[k]}] = struct{}{}
				}
			}
		}
	}
	res := [][]int{}
	for k, _ := range resMap {
		res = append(res, []int{k[0], k[1], k[2]})
	}
	return res
}




func main(){
	nums := []int{-1,0,1,2,-1,-4}
	fmt.Println(threeSum(nums))
}